File: python/a08.py
# A08 Two Dimensional Sum
# submission: https://atcoder.jp/contests/tessoku-book/submissions/38344227

h, w = map(int, input().split())
x = [[0] * (w + 1)] + [[]] * h

for i in range(h):
    x[i + 1] = [0] + list(map(int, input().split()))

for i in range(h + 1):
    for j in range(w):
        x[i][j + 1] += x[i][j]

for i in range(h):
    for j in range(w + 1):
        x[i + 1][j] += x[i][j]

q = int(input())

for qq in range(q):
    a, b, c, d = map(int, input().split())
    a -= 1
    b -= 1
    print(x[c][d] - x[a][d] - x[c][b] + x[a][b])

File: rust/tessoku-book/src/bin/a08.rs
// A08 Two Dimensional Sum
// submission: https://atcoder.jp/contests/tessoku-book/submissions/38344342

use proconio::input;

fn main() {
    input!{
        h: usize,
        w: usize,
        x:[[u32; w]; h],
        q: usize,
    }
    let mut y = vec![vec![0u32; w + 1]; h + 1];
    for i in 0..h {
        for j in 0..w {
            y[i + 1][j + 1] = y[i + 1][j] + x[i][j];
        }
    }
    for i in 0..h {
        for j in 0..w {
            y[i + 1][j + 1] += y[i][j + 1];
        }
    }
    for _qq in 0..q {
        input!{
            mut a: usize,
            mut b: usize,
            c: usize,
            d: usize,
        }
        a -= 1;
        b -= 1;
        println!("{}", y[c][d] - y[a][d] - y[c][b] + y[a][b]);
    }
}
